### PR TITLE
feat(har): re-add routeFromHAR

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: packages/playwright-core/src/client/harRouter.ts [hunk 1]
```diff
@@ -73,20 +92,29 @@ export class HarRouter {
...
-  async addRoute(context: BrowserContext) {
-    await context.route(this._pattern, route => this._handle(route));
+  async addContextRoute(context: BrowserContext) {
+    await context.route(this._options.urlMatch || '**/*', route => this._handle(route));
     context.once(Events.BrowserContext.Close, () => this.dispose());
   }
 
+  async addPageRoute(page: Page) {
+    await page.route(this._options.urlMatch || '**/*', route => this._handle(route));
+    page.once(Events.BrowserContext.Close, () => this.dispose());
```
Comment (dgozman) [inline packages/playwright-core/src/client/harRouter.ts:115]:
```suggestion
    page.once(Events.Page.Close, () => this.dispose());
```

File: packages/playwright-core/src/client/harRouter.ts [hunk 2]
```diff
@@ -14,37 +14,56 @@
...
     this._harId = harId;
-    this._pattern = options?.urlFilter ?? /.*/;
     this._options = options;
+    this._fallback = fallback;
   }
 
   private async _handle(route: Route) {
     const request = route.request();
+
+    if (this._options.urlMatch && !urlMatches(this._options.baseURL, request.url(), this._options.urlMatch)) {
```
Comment (dgozman) [inline packages/playwright-core/src/client/harRouter.ts:62]: Why do we match here, and also pass `urlMatch` to `page.route()`?

File: docs/src/api/class-browsercontext.md [hunk 3]
```diff
@@ -1025,6 +1025,29 @@ handler function to route the request.
...
+
+### option: BrowserContext.routeFromHAR.fallback
+- `fallback` <[HarFallback]<"abort"|"continue"|"fallback">>
+
+* If set to 'abort' any request not found in the HAR file will be aborted.
+* If set to 'continue' missing requests will be sent to the network.
+* If set to 'fallback' falls through to the next route handler in the handler chain.
+
+### option: BrowserContext.routeFromHAR.url
+- `url` <[string]|[RegExp]|[function]\([URL]\):[boolean]>
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply docs/src/api/class-browsercontext.md:1047]: url is consistent with the route(url) parameter.

File: docs/src/api/class-page.md [hunk 4]
```diff
@@ -2732,6 +2732,30 @@ handler function to route the request.
...
+### param: Page.routeFromHAR.har
+- `har` <[path]>
+
+Path to a [HAR](http://www.softwareishard.com/blog/har-12-spec) file with prerecorded network data. If `path` is a relative path, then it is resolved relative to the current working directory.
+
+### option: Page.routeFromHAR.fallback
+- `fallback` ?<[HarFallback]<"abort"|"fallback">>
+
+* If set to 'abort' any request not found in the HAR file will be aborted.
+* If set to 'continue' missing requests will be sent to the network.
```
Comment (dgozman) [inline docs/src/api/class-page.md:2750]:
```suggestion
8 If set to 'fallback' falls through to the next route handler in the handler chain.
```

**Review by dgozman — APPROVED** [review]

**Pull request merged**